### DISCUSSION TITLE
Adding JAVA Code Level Visibility in Skywalking

### DISCUSSION THREAD

**Discussion by AjinThankachan** [apache/skywalking#8984] [Q&A] [answered]
Hi,
I am planning to add java code-level visibility to skywalking. Which will provide more detailed visibility and ease of error finding in code levels. Is there any example of this that is already done like any portion or first level visibility kind of code. I am a bit confused on this point how to implement it. So can someone help me to do that and we have a good feature also. It would be nice if some one can tell me how to do it and what are the options available. Also if we need to add the same to .net applications is there any change in collector side? Or need to develop separate code for each agent?
Thanks in advance.

**Comment by wu-sheng** [comment, marked as answer]
What is code-level visibility? This is not a common concept, could you provide some demo?

**Reply by AjinThankachan** [reply]
Sorry for the confusion. In tracing I can see mostly the http requests only. We are not getting any application code errors or class issues,concrete class and the method invocation situation(As per my understanding) so if we can add more in-depth monitoring it would be great na. So the developers can easily debug and resolve the issues.

**Reply by wu-sheng** [reply]
HTTP only because we don't need those spans to add payload to your services. If a method is slow, profiling is ready to use. This is a basic root cause locating method in a distributed system. Always using spans is not a good practice at all.
Agent profiling is available for Java and Python, ebpf profiling is on going for C, C++ and Golang.

Although, if you insist to add more spans, then I assume you mean in Java, because you mentioned HTTP only, there are several plugins, such as, Spring annotation instrumentation and XML based instrumentation to use out of box.

**Reply by wu-sheng** [reply]
One important thing, distributed tracing in SkyWalking is for product services, which have high payload, we need to use reasonable resources to locate bottleneck, although our agents are powerful enough to get any methods' metrics, they are very expensive if you do that. We are open-source, so, you could make your own call for your env.
